### ISSUE TITLE
Implement vanilla commands

### ISSUE THREAD

**Issue by rom1504** [mhsjlw/flying-squid#126]
http://minecraft.gamepedia.com/Commands#Summary_of_commands

I need /tp

**Comment by 109C** [comment]
I think they should be implementing in a plugin separate from the main repo, but installed by default.

**Comment by rom1504** [comment]
That might be a good idea indeed since, unlike other (internal) plugins, nothing depends on commands.

**Comment by demipixel** [comment]
Hmm. From a plugin standpoint, I would expect

```js
for (var c of player.commands) {
  player.commands[c].disabled = true;
}
```

**Comment by rom1504** [comment]
would be nice to put commands in api.md (or a commands.md idk). Would make it easier to know all the commands available during dev.

**Comment by rom1504** [comment]
We need [target selector](http://minecraft.gamepedia.com/Commands#Target_selectors) too. They'll be useful for /kill and /attach

**Comment by demipixel** [comment]
just btw /attach isn't a real command...

Also, what format do you want for target selectors? For example, do we want to allow

```js
new TargetSelector({
  selector: 'entity',
  type: 'player',
  position: X,
  radius: 4
});
```
And then deal with the text processing afterwards?

**Comment by rom1504** [comment]
Yeah I know it's not really a command, but it's useful for testing, once attaching is properly implemented, /attach can go in a plugin.

Are target selectors useful in the code ? I mostly meant we need them in command parameters (as in `/kill @e`)

**Comment by demipixel** [comment]
Well, we've been using whitelist/blacklist for no (arrays of players). The only positive about a selector is that the "array" can change (hence the use of it in command blocks. Command blocks are set once but entities they affect can change).

**Comment by rom1504** [comment]
ah ok, yeah that makes sense

**Comment by demipixel** [comment]
Then again, I don't think we need it for plugins. Command blocks will be parsed each time they're run, so we just recalculate the arrays then. Whitelist/Blacklist is easier to handle.

**Comment by rom1504** [comment]
thanks @azastrael for doing the /tp commands

I think we should do a list of commands to do here with [ ] , to keep track of progress.

**Comment by rom1504** [comment]
I put the list in the first post

**Comment by demipixel** [comment]
Started selectors (incomplete) on /demipixel/flying-squid "selector" branch

**Comment by rom1504** [comment]
I added "easy" labels for some commands that are pretty easy (for example because they don't require any new flying-squid feature)

I think this is a good starting point for new contributors.

**Comment by rom1504** [comment]
I added tests for a good portion of the currently implemented commands.
These commands tests are pretty easy to write and understand (see https://github.com/mhsjlw/flying-squid/blob/master/test/mineflayer.js#L93).

For new added commands, it would be nice to add such a test with it.

**Comment by rom1504** [comment]
selectors are mostly done (by @demipixel ), now we can use them in commands :)

**Comment by mhsjlw** [comment]
@rom1504 should this be in the TODO.md?

**Comment by rom1504** [comment]
no, issues are better really.
TODO.md is useful for a very general roadmap maybe (like currently)

**Comment by mhsjlw** [comment]
Ok

**Comment by demipixel** [comment]
I think /me and /say would be easy.

Also, if you're not new to the project, don't do the easy ones. Thanks.

**Comment by rom1504** [comment]
It would be nice to add an explanation into how to add a new command in the api.md. Take an example (like https://github.com/PrismarineJS/flying-squid/blob/8c4406804e015054ccd81568652ece42fac819a1/src/lib/plugins/spawn.js#L156 ) and add comments around it to explain what is what.

**Comment by rom1504** [comment]
done in 6040a4a12d4e15b5611bca666c1a4a38089cc709

**Comment by andriycraft** [comment]
Remove /publish command because it avalible only in singleplayer

**Comment by AndyIsHereBoi** [comment]
Is there a updated table of what commands need to be added?
